### PR TITLE
[SLP]Better cost estimation for masked gather or "clustered" loads.

### PR DESCRIPTION
**Pull request by alexey-bataev**
After landing support for actual vectorization of the "clustered" loads,
need better estimate the cost between the masked gather and clustered loads.
This includes estimation of the address calculation and better
estimation of the gathered loads. Also, this estimation now relies on
SLPCostThreshold option, allowing modify the behavior of the compiler.

### CODE REVIEW

**Comment by alexey-bataev** [comment]
Ping!

**Comment by alexey-bataev** [comment]
Ping!

**Comment by alexey-bataev** [comment]
@RKSimon Simon, could you check this one, please?

**Review by RKSimon — APPROVED** [review]
LGTM with one minor

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -4790,105 +4790,174 @@ BoUpSLP::LoadsState BoUpSLP::canVectorizeLoads(
...
+    // Estimate the cost of masked gather GEP. If not a splat, roughly
+    // estimate as a buildvector, otherwise estimate as splat.
+    if (static_cast<unsigned>(count_if(
+            PointerOps, IsaPred<GetElementPtrInst>)) < PointerOps.size() - 1 ||
+        any_of(PointerOps, [&](Value *V) {
+          return getUnderlyingObject(V) !=
+                 getUnderlyingObject(PointerOps.front());
+        }))
+      VectorGEPCost += TTI.getScalarizationOverhead(
+          VecTy, APInt::getAllOnes(VecTy->getElementCount().getKnownMinValue()),
```
[the remark this answers is not in the document]
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:4812]: Yep, will fix

**Pull request merged**